### PR TITLE
Add grpc-service-reactor example

### PR DESCRIPTION
**Pull request by cbornet**
This adds an example for gRPC using the [reactive-grpc](https://github.com/salesforce/reactive-grpc) library.

See https://github.com/line/armeria-examples/pull/2

I, @cbornet, hereby release the Git commits in [this pull request](https://github.com/line/armeria-examples/pull/2310) ('The Contribution') under [Creative Commons Zero 1.0](https://creativecommons.org/publicdomain/zero/1.0/), so that anyone can modify and/or redistribute The Contribution without any restriction.

### CODE REVIEW

File: examples/grpc-service-reactor/src/main/java/example/armeria/grpc/reactor/HelloServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    /**
+     * Sends a {@link HelloReply} 3 seconds after receiving a request.
+     */
+    @Override
+    public Mono<HelloReply> lazyHello(Mono<HelloRequest> request) {
+        // You can use the event loop for scheduling a task.
+        return request
+            .delayElement(Duration.ofSeconds(3), Schedulers.fromExecutor(ServiceRequestContext.current().contextAwareEventLoop()))
```
Comment (trustin) [inline examples/grpc-service-reactor/src/main/java/example/armeria/grpc/reactor/HelloServiceImpl.java]: Maybe better breaking the line somewhere?

File: examples/grpc-service-reactor/src/test/java/example/armeria/grpc/reactor/HelloServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        }
+    }
+
+    private static String uri() {
+        return "gproto+http://127.0.0.1:" + server.activeLocalPort() + '/';
+    }
+
+    @Test
+    public void getReply() {
+        final HelloServiceGrpc.HelloServiceBlockingStub helloService = Clients.newClient(uri(), HelloServiceGrpc.HelloServiceBlockingStub.class);
```
Comment (trustin) [inline examples/grpc-service-reactor/src/test/java/example/armeria/grpc/reactor/HelloServiceTest.java]: nit: Consider breaking the line after ` = `.

File: examples/grpc-service-reactor/src/test/java/example/armeria/grpc/reactor/HelloServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        assertThat(helloService.hello(HelloRequest.newBuilder().setName("Armeria").build()).getMessage())
+                .isEqualTo("Hello, Armeria!");
+    }
+
+    @Test
+    public void getReplyWithDelay() {
+        final HelloServiceGrpc.HelloServiceFutureStub helloService = Clients.newClient(uri(), HelloServiceGrpc.HelloServiceFutureStub.class);
```
Comment (trustin) [inline examples/grpc-service-reactor/src/test/java/example/armeria/grpc/reactor/HelloServiceTest.java]: nit: Consider breaking the line after ` = `.

File: examples/grpc-service-reactor/src/test/java/example/armeria/grpc/reactor/HelloServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+                throw new Error(t);
+            }
+        }, MoreExecutors.directExecutor());
+
+        await().untilTrue(completed);
+    }
+
+    @Test
+    public void getReplyFromServerSideBlockingCall() {
+        final HelloServiceGrpc.HelloServiceBlockingStub helloService = Clients.newClient(uri(), HelloServiceGrpc.HelloServiceBlockingStub.class);
```
Comment (trustin) [inline examples/grpc-service-reactor/src/test/java/example/armeria/grpc/reactor/HelloServiceTest.java]: nit: Consider breaking the line after ` = `.

**Comment by trustin** [comment]
Because the directory you are contributing to is intentionally licensed under CC0, you need to explicitly comment that you release your contribution in this pull request under Creative Commons Zero 1.0 - https://creativecommons.org/publicdomain/zero/1.0/

For example:

    I, @cbornet, hereby release the Git commits in 
    [this pull request](https://github.com/line/armeria-examples/pull/2) 
    ('The Contribution') under 
    [Creative Commons Zero 1.0](https://creativecommons.org/publicdomain/zero/1.0/), 
    so that anyone can modify and/or redistribute The Contribution without 
    any restriction.

@syleeeee What do you think?

**Comment by trustin** [comment]
I just realized `/settings.gradle` was not updated. Could you add the new example there so that it's built as a part of the project? (Please don't forget to rebase on or merge master, because I pushed a change which modifies `/settings.gradle` to master just today.)

**Comment by cbornet** [comment]
> Thanks! Is there a reason we need to keep the resources directories in git? If not we can leave them out.

It's a direct copy of the grpc-service folder. I can remove them of course.

**Comment by anuraaga** [comment]
I didn't notice it's a copy - don't worry about it :)

**Comment by cbornet** [comment]
> Because the directory you are contributing to is intentionally licensed under CC0, you need to explicitly comment that you release your contribution in this pull request under Creative Commons Zero 1.0 - https://creativecommons.org/publicdomain/zero/1.0/

This one here is Apache 2, isn't it ?

**Comment by trustin** [comment]
> This one here is Apache 2, isn't it ?

Not the files under the `examples` directory.

File: examples/README.md [hunk 5]
```diff
@@ -9,6 +9,12 @@
   - See [Running a gRPC service](https://line.github.io/armeria/server-grpc.html) and
     [Calling a gRPC service](https://line.github.io/armeria/client-grpc.html).
 
+- `grpc-service-reactor`
+  - Learn how to write a gRPC service with Armeria gRPC module, [reactive-grpc](https://github.com/salesforce/reactive-grpc) and [reactor](https://projectreactor.io/) libraries for
```
Comment (trustin) [inline examples/README.md]: How about:

    [`reactive-grpc`](...) and [`reactor`](...)

..?

Comment (cbornet) [reply examples/README.md]: done

File: examples/grpc-service-reactor/src/main/java/example/armeria/grpc/reactor/HelloServiceImpl.java (hunk 1 above)
Comment (cbornet) [reply examples/grpc-service-reactor/src/main/java/example/armeria/grpc/reactor/HelloServiceImpl.java]: done

File: examples/grpc-service-reactor/src/test/java/example/armeria/grpc/reactor/HelloServiceTest.java (hunk 2 above)
Comment (cbornet) [reply examples/grpc-service-reactor/src/test/java/example/armeria/grpc/reactor/HelloServiceTest.java]: done

File: examples/grpc-service-reactor/src/test/java/example/armeria/grpc/reactor/HelloServiceTest.java (hunk 3 above)
Comment (cbornet) [reply examples/grpc-service-reactor/src/test/java/example/armeria/grpc/reactor/HelloServiceTest.java]: done

File: examples/grpc-service-reactor/src/test/java/example/armeria/grpc/reactor/HelloServiceTest.java (hunk 4 above)
Comment (cbornet) [reply examples/grpc-service-reactor/src/test/java/example/armeria/grpc/reactor/HelloServiceTest.java]: done

**Comment by cbornet** [comment]
@trustin done.

**Comment by trustin** [comment]
Ah, you did already. Thanks a lot!

**Comment by syleeeee** [comment]
> Because the directory you are contributing to is intentionally licensed under CC0, you need to explicitly comment that you release your contribution in this pull request under Creative Commons Zero 1.0 - https://creativecommons.org/publicdomain/zero/1.0/
> 
> For example:
> 
> ```
> I, @cbornet, hereby release the Git commits in 
> [this pull request](https://github.com/line/armeria-examples/pull/2) 
> ('The Contribution') under 
> [Creative Commons Zero 1.0](https://creativecommons.org/publicdomain/zero/1.0/), 
> so that anyone can modify and/or redistribute The Contribution without 
> any restriction.
> ```
> 
> @syleeeee What do you think?

Guess I'm late but it looks fine. Thanks for the example @trustin . :) 

And thanks for your contribution too, @cbornet . 👍

File: gradle/scripts/lib/java-rpc-proto.gradle [hunk 6]
```diff
@@ -25,12 +25,20 @@ configure(projectsWithFlags('java')) {
                     grpc {
                         artifact = "io.grpc:protoc-gen-grpc-java:${managedVersions['io.grpc:grpc-core']}"
                     }
+                    if (project.ext.hasFlag('reactor-grpc') && managedVersions.containsKey('com.salesforce.servicelibs:reactor-grpc')) {
```
[the remark this answers is not in the document]
Comment (trustin) [reply gradle/scripts/lib/java-rpc-proto.gradle]: Yep, I will.

**Comment by trustin** [comment]
Thanks a again for your example, @cbornet. It's a nice addition that fills the gap! :bowing_man:

You also might be interested in writing an example that uses `rxgrpc` by the way. :wink:

**Comment by trustin** [comment]
Added a follow-up commit: a2ebe29bb098185046d7741e603ee96d04b0c55b

**Comment by cbornet** [comment]
Reactive grpc should work with Gradle on Windows : https://github.com/salesforce/reactive-grpc/issues/186

**Comment by trustin** [comment]
I didn't test it by myself, but just read that from the official `README.md`. Perhaps it's outdated? Let me revert it, then. Thanks for letting me know.

**Comment by cbornet** [comment]
I think it's indeed outdated. @rmichela wdyt ?

**Comment by rmichela** [comment]
Gradle on Windows is supported. Docs are out of date.

**Comment by cbornet** [comment]
Thanks Ryan. I'll update the docs then.

**Pull request merged**